core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableInterpretable.java

```diff
@@ -82,6 +89,16 @@ public Node implement(final InterpreterImplementor implementor) {
     return new EnumerableNode(enumerable, implementor.compiler, this);
   }
 
+  private static final int BINDABLE_CACHE_MAX_SIZE =
+      Util.getIntProperty("calcite.bindable.cache.maxSize", (size) -> size >= 0, 0);
+  private static final int BINDABLE_CACHE_CONCURRENCY_LEVEL =
+      Util.getIntProperty("calcite.bindable.cache.concurrencyLevel", (level) -> level > 0, 1);
+  private static final Cache<String, Bindable> BINDABLE_CACHE =
```

Comment:
@zabetak there is no way to make `BINDABLE_CACHE` instance this non-static ?

Reply:
We want to cache `Bindable` objects across connections so I don't see a relevant class which can hold the cache. Moreover, having the cache stored in another object would mean that we need to pass it using parameters till `getBindable` (which by the way is static) and that would introduce many changes. For the previous reasons, I would rather keep it as it is but if you see a better option, I am willing to follow:)

Reply:
Agree that removing statics can be complicated. 
Just for my understanding, why not have cache per root schema (connection) ?

Reply:
It could be associated with the schema. However, a schema (the same model) can come and go multiple times (if we make multiple connections or other similar stuff) so it is a pity to lose queries that are already in the cache. Having said that the main reason for leaving it static is to avoid changing a whole bunch of existing methods.

Reply:
What would happen if we initialize two connections with different `config.getInSubQueryThreshold()` will that use previously cached query ? 

If I understand correctly queries can be different depending on `SqlToRelConverter.Config`.

Reply:
A valid question.

But just don't do that.

We can't solve it without answering the question "Is calcite a library or a server?". And I don't think we want to pick one. So we're going to have to live with inconsistency.

Reply:
> Is calcite a library or a server?

*When possible* code as a library. Since it is easier to go from library to server than the other way around.

Reply:
But you see my point, right? People can't use it as a library if some of the configuration, such as this, is global.

If we allow use as a library and a server, which we do, there are going to be inconsistencies.

Reply:
> But you see my point, right?
Yes, I do. That's why global variables (except constants) are a red flag to me.

Reply:
On reflection maybe global configuration is not so bad (it's OK if all users of the library have the same bindable cache size) but a shared cache is really bad.

Reply:
Just to play devil's advocate. 
If you allow separate caches they why not allow different settings for those caches ?

Reply:
Thanks for your time and the comments @asereda-gs , @julianhyde . 

I would like to highlight a few points.
1. The cache in this class is rather low level since we are storing the java code it self (as a string) and not an SQL query or a plan. This is not memory efficient but solves the main performance penalty of class loading (JIT etc.) without big risk of side effects.
2. The cache was added at this level to benefit clients using Calcite as a library and as a server. Clients can benefit from it either by passing through CalcitePrepare and also by calling directly `EnumerableInterpretable#toBindable`.
3. The most important performance gain comes from queries which are executed frequently (consecutively) over a very short period of time. This cache is simple and at the same time solves the problem for the most common appearing patterns. 

For the reasons outlined above I do not see a big benefit having separate caches per schema, connection, or something else but I may have underestimated some aspect. 

Correct me if I am wrong but from your comments I get the impression that you are not in favor of merging this PR. I would like to include it in 1.19 but obviously I will not proceed if you are not ok with it.

Reply:
I am in favor of merging this PR.

I think that the objections that @asereda-gs is raising are a bit theoretical, and we should not let them stop us.

To @asereda-gs's specific question:

> Just to play devil's advocate.
> If you allow separate caches they why not allow different settings for those caches ?
> This is especially true during testing.

The reason to not allow more fine-grained configuration is that it adds complexity, and it is complexity that no one is asking for.

Reply:
I'm also in favor of this PR. However not yet 100% sure that cache should exists globally (as static variable) or at connection level (root schema). 

Probably one can review this decision later-on and it shouldn't prevent merging current proposal.

Reply:
OK great, thanks for your input! I will try to come up with a better name for `calcite.bindable.cache.maxSize` and merge this PR the following days.